datafusion/core/src/datasource/physical_plan/parquet/statistics.rs

```diff
@@ -552,6 +552,22 @@ make_data_page_stats_iterator!(MinInt32DataPageStatsIterator, min, Index::INT32,
 make_data_page_stats_iterator!(MaxInt32DataPageStatsIterator, max, Index::INT32, i32);
 make_data_page_stats_iterator!(MinInt64DataPageStatsIterator, min, Index::INT64, i64);
 make_data_page_stats_iterator!(MaxInt64DataPageStatsIterator, max, Index::INT64, i64);
+make_data_page_stats_iterator!(
```

Comment:
@tshauck 
It seems like you are not using those in the maching down below? that might cause your error on f16. There you are using an Float32 iterator which won't match on Index::FIXED_LEN_BYTE_ARRAY and thus return None.

I think you need to switch to the f16 iterator; but perhaps might need to construct an f16 from the bytes.

Reply:
Apologies for the confusion. It pushed this as a WIP. I still need to work through the f16 case. I think I will need to construct an f16 from the bytes, but not sure yet.

Reply:
Here's an example of what I mean...

<img width="1130" alt="image" src="https://github.com/apache/datafusion/assets/421839/852f9b91-7b50-4903-8c59-cdd0bcd15fca">

Fixed length byte array is collected as `parquet::data_type::FixedLenByteArray`, which then means this doesn't match. Alternatively, using that data type in the macro causes later move issues when interacting with the calls to `min()` and `max()`.

Reply:
I suggest we take a friendly look at how f16 for row groups work: https://github.com/apache/datafusion/blob/1cb0057b99a016ee982d34deb860cf7f4e52b845/datafusion/core/src/datasource/physical_plan/parquet/statistics.rs#L336-L340

I suspect the statistics in data pages will be encodd the same

It is also possible that since f16 is a less common type, the parquet writer might write data page statistics for it

Reply:
hi there, I haven't noticed this PR and accidentally addressed the same issue in here https://github.com/apache/datafusion/pull/10982 . Apologies. I think the f32 and f64 cases we solve the ~same. I got the f16 working, though not in a way I'm particularly happy with (ie, the move issues you mention)

I don't have any other commits in that PR in my queue -- so happy to close it if you come up with any other f16 solution...

Reply:
Oooff, sorry @tmi, I should've done `take` on the issue. Taking @alamb's pointer to look at row groups, I might be wrong, but I think there's a couple of differences between it and the data page stats that complicate things...

* `max_bytes`/`min_bytes`, which is used for row groups, does not exist on `PageIndex` in the parquet crate but does on `Statistics`, so using it in the macro isn't possible at this point.
* Using `[u8]` as the `stat_value_type` is problematic when combined with the iterator macro in that `type Item = Vec<Option<$stat_value_type>>;` can't work with `[u8]` because it's unsized. Changing the item def to `type Item = Vec<Option<&'a $stat_value_type>>;` works and is similar to the item definition of page stats (`type Item = Option<&'a $stat_value_type>;`).

f16 is less common, so maybe to have a solution that's similar to Statistics and hopefully avoids cloning, we could:

1. Update this PR to do f32 and f64, get it in to support higher priority types and is mostly innocuous
2. Update the parquet to support `max_bytes`/`min_bytes` on `PageIndex`
3. Update the data page stats to use a `&'a $stat_value_type` and add f16 when that's merged

@tmi, maybe you and I could collaborate on 2 & 3 if that seems reasonable? Otherwise, certainly happy to go another route, if there's a better approach.

Reply:
Sounds like a good plan to me

I think it is perfectly reasonable to leave `f16` support as a 'TODO' in datafusion as we add the needed support upstream

Let me know if I can help (e.g. file some tickets, etc

Reply:
I think https://github.com/apache/datafusion/pull/10982 is good to go FWIW

Reply:
Ok, sounds good to me. Maybe a further improvement down the line to avoid the copy and simplify to use a method vs a function. Shall I close this one?

Reply:
> avoid the clone

I think the ideal sequence would be
1. https://github.com/rust-lang/rust/issues/116909 gets merged
2. parquet replaces FixedLengthByteArray with genuine f16 type
3. we drop the special handling here and elsewhere

and until then the f16 support in datafusion would be as is... Or alternatively, introduce the half crate to parquet rs first. Because dealing with FixedLengthByteArray on the user side is always going to be messy